feat(aws-pcs): cluster-scoped Name tag + PCS-API login lookup in docs (#1183)
aws-samples/awsome-distributed-training: architectures/aws-pcs/README.md

on the login node to that CIDR.

**Console:** [EC2 Console](https://console.aws.amazon.com/ec2/home#Instances:) → filter
by `Name` = `<your-stack-name>-login` → **Connect** → **Session Manager**.

**CLI** (AWS CloudShell has the required permissions). Set `STACK_NAME` /
`AWS_REGION` once, then a single command resolves the login node's
instance ID — it works no matter what your CNG is named because the
lookup goes through the PCS API, not the human-facing tag:

```bash
STACK_NAME=pcs-ml-cluster        # your CloudFormation stack name
AWS_REGION=us-east-1             # your region

> LOGIN_INSTANCE_ID=$(aws ec2 describe-instances --region "$AWS_REGION" --filters "Name=tag:aws:pcs:compute-node-group-id,Values=$(aws pcs list-compute-node-groups --cluster-identifier $(aws cloudformation describe-stacks --stack-name "$STACK_NAME" --region "$AWS_REGION" --query 'Stacks[0].Outputs[?OutputKey==`ClusterId`].OutputValue' --output text) --region "$AWS_REGION" --query 'computeNodeGroups[?name==`login`].id' --output text)" "Name=instance-state-name,Values=running" --query 'Reservations[0].Instances[0].InstanceId' --output text)

Reviewer: ### The nested `LOGIN_INSTANCE_ID=$(…)` one-liner fails silently to `None`, and `--cluster-identifier` is unquoted — 🟢 [confirmed]
**Observation:** The triple-nested substitution (this line in §6, and the §8.2 copy) is functionally correct — the JMESPath backticks are protected by single quotes, and nested `$(…)` parses fine. But there's no error path: a wrong `STACK_NAME`/`AWS_REGION` or a missing `ClusterId` output makes the inner `describe-stacks` return empty → `list-compute-node-groups` runs with an empty identifier → the `Values=` filter matches nothing → `--output text` prints literal **`None`** → `aws ssm start-session --target None` fails opaquely. Separately, `--cluster-identifier $(…)` is **unquoted**, so on an empty ClusterId the token vanishes and the following `--region` is consumed as the identifier value — a different, more confusing error than a clean "empty identifier."
**Impact:** A first-time copy-paste into the wrong stack/region gets `None` with no hint why. (Test 15's "no `None`/empty" gate is exactly the right guard — but the snippet itself gives the user no diagnostic.)
**Suggestion:** Split into named steps with a guard, and quote the identifier — IAM.md already splits out `CLUSTER_ID` for the Grafana step, so the pattern exists:
```bash
CLUSTER_ID=$(aws cloudformation describe-stacks --stack-name "$STACK_NAME" --region "$AWS_REGION" \
  --query 'Stacks[0].Outputs[?OutputKey==`ClusterId`].OutputValue' --output text)
[ -n "$CLUSTER_ID" ] && [ "$CLUSTER_ID" != "None" ] || { echo "No ClusterId — check STACK_NAME/AWS_REGION"; return 1; }
CNG_ID=$(aws pcs list-compute-node-groups --cluster-identifier "$CLUSTER_ID" --region "$AWS_REGION" \
  --query 'computeNodeGroups[?name==`login`].id' --output text)
LOGIN_INSTANCE_ID=$(aws ec2 describe-instances --region "$AWS_REGION" \
  --filters "Name=tag:aws:pcs:compute-node-group-id,Values=$CNG_ID" "Name=instance-state-name,Values=running" \
  --query 'Reservations[0].Instances[0].InstanceId' --output text)
```

on the login node to that CIDR.

**Console:** [EC2 Console](https://console.aws.amazon.com/ec2/home#Instances:) → filter
by `Name` = `<your-stack-name>-login` → **Connect** → **Session Manager**.

**CLI** (AWS CloudShell has the required permissions). Set `STACK_NAME` /
`AWS_REGION` once, then a single command resolves the login node's

> instance ID — it works no matter what your CNG is named because the

Reviewer: ### §6 rationale is slightly self-contradictory — the resilience is against the `Name` tag, not the CNG name

Small wording thing: the §6 intro says the lookup "works no matter what your CNG is named because the lookup goes through the PCS API." But the lookup hardcodes `computeNodeGroups[?name==\`login\`]`, so it *does* depend on the login CNG being named `login` (which deploy-all guarantees). What the PCS-API path actually buys you is independence from the mutable EC2 `Name` tag — that's the real win worth stating.

```suggestion
instance ID — it works no matter how instances are Name-tagged because the
```

lookup goes through the PCS API, not the human-facing tag:

```bash
STACK_NAME=pcs-ml-cluster        # your CloudFormation stack name
AWS_REGION=us-east-1             # your region

CLUSTER_ID=$(aws cloudformation describe-stacks --stack-name "$STACK_NAME" --region "$AWS_REGION" --query 'Stacks[0].Outputs[?OutputKey==`ClusterId`].OutputValue' --output text)
[ -n "$CLUSTER_ID" ] && [ "$CLUSTER_ID" != "None" ] || { echo "No ClusterId — check STACK_NAME/AWS_REGION"; return 1; }

LOGIN_CNG_ID=$(aws pcs list-compute-node-groups --cluster-identifier "$CLUSTER_ID" --region "$AWS_REGION" --query 'computeNodeGroups[?name==`login`].id' --output text)

> LOGIN_INSTANCE_ID=$(aws ec2 describe-instances --region "$AWS_REGION" --filters "Name=tag:aws:pcs:compute-node-group-id,Values=$LOGIN_CNG_ID" "Name=instance-state-name,Values=running" --query 'Reservations[0].Instances[0].InstanceId' --output text)

Reviewer: ### The split login-lookup guards only the first of three hops — a renamed `login` CNG still lands on `--target None`

The split + `CLUSTER_ID` guard is a solid fix for the round-1 silent-`None`. The residual: the guard covers `CLUSTER_ID`, but `LOGIN_CNG_ID` and `LOGIN_INSTANCE_ID` are unguarded. If `list-compute-node-groups` finds no CNG named `login` (renamed, or not yet running), `LOGIN_CNG_ID` comes back empty → the `describe-instances` filter becomes `Values=` → `LOGIN_INSTANCE_ID=None` → `aws ssm start-session --target None` fails with an opaque SSM error — the round-1 class, one hop down. A single guard right before the session covers all three hops:

```bash
[ -n "$LOGIN_INSTANCE_ID" ] && [ "$LOGIN_INSTANCE_ID" != "None" ] || { echo "No running login node for cluster $CLUSTER_ID"; return 1; }
```

Two asides: the existing `CLUSTER_ID` guard uses `return 1`, which only halts inside a function or a sourced script — pasted at an interactive prompt (which the README instructs) `return` prints an error and execution *continues* anyway, so the guard's "stop here" intent isn't enforced interactively. And credit due — the new Test 15 gate (`echo $LOGIN_INSTANCE_ID` must print an `i-…`) already catches this before a customer would, so this is polish, not a hole. `docs/IAM.md`'s equivalent block has the same shape and would take the same one-line guard.
